include/aws/common/common.h

```diff
@@ -115,30 +115,41 @@
-#define AWS_POSTCONDITION_2(cond, explanation) __CPROVER_assert((cond), (explanation))
-#define AWS_POSTCONDITION_1(cond) __CPROVER_assert((cond), #cond " check failed")
-#define AWS_MEM_IS_READABLE(base, len) __CPROVER_r_ok((base), (len))
-#define AWS_MEM_IS_WRITABLE(base, len) __CPROVER_w_ok((base), (len))
+#    define AWS_PRECONDITION_2(cond, explanation) __CPROVER_precondition((cond), (explanation))
+#    define AWS_PRECONDITION_1(cond) __CPROVER_precondition((cond), #cond " check failed")
+#    define AWS_FATAL_PRECONDITION_2(cond, explanation) __CPROVER_precondition((cond), (explanation))
+#    define AWS_FATAL_PRECONDITION_1(cond) __CPROVER_precondition((cond), #cond " check failed")
+#    define AWS_POSTCONDITION_2(cond, explanation) __CPROVER_assert((cond), (explanation))
+#    define AWS_POSTCONDITION_1(cond) __CPROVER_assert((cond), #cond " check failed")
+#    define AWS_FATAL_POSTCONDITION_2(cond, explanation) __CPROVER_assert((cond), (explanation))
+#    define AWS_FATAL_POSTCONDITION_1(cond) __CPROVER_assert((cond), #cond " check failed")
+#    define AWS_MEM_IS_READABLE(base, len) __CPROVER_r_ok((base), (len))
+#    define AWS_MEM_IS_WRITABLE(base, len) __CPROVER_w_ok((base), (len))
 #else
-#define AWS_PRECONDITION_2(cond, expl) AWS_ASSERT(cond)
-#define AWS_PRECONDITION_1(cond) AWS_ASSERT(cond)
-#define AWS_POSTCONDITION_2(cond, expl) AWS_ASSERT(cond)
-#define AWS_POSTCONDITION_1(cond) AWS_ASSERT(cond)
+#    define AWS_PRECONDITION_2(cond, expl) AWS_ASSERT(cond)
+#    define AWS_PRECONDITION_1(cond) AWS_ASSERT(cond)
+#    define AWS_FATAL_PRECONDITION_2(cond, expl) AWS_FATAL_ASSERT(cond)
+#    define AWS_FATAL_PRECONDITION_1(cond) AWS_FATAL_ASSERT(cond)
+#    define AWS_POSTCONDITION_2(cond, expl) AWS_ASSERT(cond)
+#    define AWS_POSTCONDITION_1(cond) AWS_ASSERT(cond)
+#    define AWS_FATAL_POSTCONDITION_2(cond, expl) AWS_FATAL_ASSERT(cond)
+#    define AWS_FATAL_POSTCONDITION_1(cond) AWS_FATAL_ASSERT(cond)
+
 /* the C runtime does not give a way to check these properties,
  * but we can at least check that the pointer is valid */
-#define AWS_MEM_IS_READABLE(base, len) (((len) == 0) || (base))
-#define AWS_MEM_IS_WRITABLE(base, len) (((len) == 0) || (base))
+#    define AWS_MEM_IS_READABLE(base, len) (((len) == 0) || (base))
+#    define AWS_MEM_IS_WRITABLE(base, len) (((len) == 0) || (base))
 #endif
 
 // The UNUSED is used to silence the complains of GCC for zero arguments in variadic macro
 #define AWS_PRECONDITION(...) GET_MACRO(__VA_ARGS__, AWS_PRECONDITION_2, AWS_PRECONDITION_1, UNUSED)(__VA_ARGS__)
+#define AWS_FATAL_PRECONDITION(...) GET_MACRO(__VA_ARGS__, AWS_FATAL_PRECONDITION_2, AWS_FATAL_PRECONDITION_1, UNUSED)(__VA_ARGS__)
 #define AWS_POSTCONDITION(...) GET_MACRO(__VA_ARGS__, AWS_POSTCONDITION_2, AWS_POSTCONDITION_1, UNUSED)(__VA_ARGS__)
+#define AWS_FATAL_POSTCONDITION(...) GET_MACRO(__VA_ARGS__, AWS_FATAL_POSTCONDITION_2, AWS_FATAL_POSTCONDITION_1, UNUSED)(__VA_ARGS__)
```

Comment:
Since you're changing these macros, could you also add `AWS_ERROR_PRECONDITION` and `AWS_ERROR_POSTCONDITION` macros that would raise error codes when the condition is violated? Something like...

```
#define AWS_ERROR_PRECONDITION(cond, err)       \
    do {                                        \
        if (cond) {                             \
            return aws_raise_error(err);        \
        }                                       \
    } while (0)
```
@danielsn @ColdenCullen what do you think?

Reply:
THIS IS COOL. I think it can be done in a separate PR tho, let's get just this in first.

Reply:
Glad you like it.  Working on a PR now.

Also, for consistency, should I rename the current one that asserts ```AWS_ASSERT_PRECONDITION```? Or just leave it as is?

Reply:
I don't feel strongly either way? I don't mind the idea of all 3 being different, but I also like that ASSERT is the default. @JonathanHenson @graebm @justinboswell any thoughts?